### ISSUE TITLE
Set up a private GitHub runner for arm64 Linux

### ISSUE THREAD

**Issue by britzl** [defold/defold#9929]
We need arm64 Linux runners on GitHub to build and test arm64 Linux support. GH is supposed to make these available by the end of the year:

_"These runners are available to our customers on our GitHub Team and Enterprise Cloud plans. We expect to begin offering Arm runners for open source projects by the end of the year."_

https://github.blog/news-insights/product-news/arm64-on-github-actions-powering-faster-more-efficient-build-systems/#get-started-using-arm-hosted-runners-today

We need to set up a private runner if this is delayed longer than Q1 2025.

**Comment by aglitchman** [comment]
It seems to me that Github will make money on enterprise customers for as long as possible, so it's not worth waiting for arm64 runners for open source this year.

My suggestion is to temporarily build in x86-64 runner via cross-compilation and run tests via qemu. I think I even managed to do it - will the team be interested in this? I'll be ready to do a PR soon.

**Comment by britzl** [comment]
> My suggestion is to temporarily build in x86-64 runner via cross-compilation and run tests via qemu. I think I even managed to do it - will the team be interested in this? I'll be ready to do a PR soon.

Yes, this is of interest to us. The alternative, which we are also considering is to use a self hosted arm64 linux runner.

@JCash and @ekharkunov what are your thoughts on this?

**Comment by ekharkunov** [comment]
> > My suggestion is to temporarily build in x86-64 runner via cross-compilation and run tests via qemu. I think I even managed to do it - will the team be interested in this? I'll be ready to do a PR soon.
> 
> Yes, this is of interest to us. The alternative, which we are also considering is to use a self hosted arm64 linux runner.
> 
> @JCash and @ekharkunov what are your thoughts on this?

From my side I suggest 2 options:
1. Use macos arm64 runners and docker container. Run engine build inside that container. Potential problem: some issues because of Docker
2. Setup self-hosted arm64 runner and configure it.Potential problems: maintenance, security concerns (https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners#self-hosted-runner-security), additional costs.

From my point of view cross-compilation is "last stand" for that task.
Personally, I'd like to start with option 2, than 1 and in the end - cross-compilation.

**Comment by JCash** [comment]
When doing this, I was looking into the cross-compilation.
But as I was time constrained, I didn't have time to debug or do trial-and-error for the CI.

As the cross compilation is already supported by our tool chains, given a working compiler (we have clang-16), and a set of libraries (the r&d I skipped), it should be doable.

If there is a PR for that, I think that's our most preferrable way.

Re macOS + Docker: I'd rather have a linux machine to run the tests on, as opposed to a VM (Docker) as we get further from the end user experience.

Re custom runners: We've talked about this, and it may be a viable approach later on, but I don't think it should be necessary for this task.

**Comment by aglitchman** [comment]
> In that case we need investigate how to run arm64 application at x86_64 (something like Rosseta?)

As I pointed above, the tests greatly runs via `qemu`. It's like Rosetta 😀

**Comment by aglitchman** [comment]
Thank you all for the lively discussion! Then I'll make a PR with the changes and you decide if it suits or not. I like this variant because I can run tests on my PC and build a version of the engine for arm64 as well. Very convenient!

**Comment by AGulev** [comment]
I think, it's outdated now?
